Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,342 @@
+
+/-- A group `G` acts properly on a topological space `X` if and only if for all ultrafilters`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `g • x₂ = x₁` and `𝒰.fst` converges to `g`. -/
+@[to_additive "A group acts `G` properly on a topological space `X` if and only if
+for all ultrafilters `𝒰` on `X`, if `𝒰` converges to `(x₁, x₂)`
+along the map `(g, x) ↦ (g • x, x)`, then there exists `g : G` such that `g • x₂ = x₁`
+and `𝒰.fst` converges to `g`."]
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto : ProperSMul G X ↔ ContinuousSMul G X
+    ∧ (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, g • x₂ = x₁ ∧ Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  constructor
+  · intro h
+    refine ⟨by infer_instance, fun 𝒰 x₁ x₂ h' ↦ ?_⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter] at h
+    have ⟨(g, x), hgx1, hgx2⟩ := h.2 h'
+    use g
+    constructor
+    · simp at hgx1
+      rw [← hgx1.2, hgx1.1]
+    · have := continuous_fst.tendsto (g, x)
+      rw [Tendsto] at *
+      calc
+        map Prod.fst ↑𝒰 ≤ map Prod.fst (𝓝 (g, x)) := map_mono hgx2
+        _               ≤ 𝓝 (g, x).1 := this
+  · rintro ⟨cont, h⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter]
+    refine ⟨by fun_prop, fun 𝒰 (x₁, x₂) hxx ↦ ?_⟩
+    rcases h 𝒰 x₁ x₂ hxx with ⟨g, hg1, hg2⟩
+    use (g, x₂)
+    refine ⟨by rw [hg1], ?_⟩
+    rw [nhds_prod_eq, 𝒰.le_prod]
+    refine ⟨hg2, ?_⟩
+    change Tendsto (Prod.snd ∘ (fun gx : G × X ↦ (gx.1 • gx.2, gx.2))) ↑𝒰 (𝓝 (Prod.snd (x₁, x₂)))
+    apply Filter.Tendsto.comp
+    apply Continuous.tendsto
+    exact continuous_snd
+    assumption
+
+/-- A group `G` acts properly on a T2 topological space `X` if and only if for all ultrafilters
+`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `𝒰.fst` converges to `g`. -/
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto_t2 [T2Space X] : ProperSMul G X ↔
+    ContinuousSMul G X ∧
+    (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  constructor
+  · intro h
+    have := properSMul_iff_continuousSMul_ultrafilter_tendsto.1 h
+    refine ⟨this.1, fun 𝒰 x₁ x₂ h' ↦ ?_⟩
+    rcases this.2 𝒰 x₁ x₂ h' with ⟨g, _, hg⟩
+    exact ⟨g, hg⟩
+  · rintro ⟨cont, h⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter_of_t2]
+    refine ⟨by fun_prop, fun 𝒰 (x₁, x₂) hxx ↦ ?_⟩
+    rcases h 𝒰 x₁ x₂ hxx with ⟨g, hg⟩
+    use (g, x₂)
+    rw [nhds_prod_eq, 𝒰.le_prod]
+    refine ⟨by assumption, ?_⟩
+    change Tendsto (Prod.snd ∘ (fun gx : G × X ↦ (gx.1 • gx.2, gx.2))) ↑𝒰 (𝓝 (Prod.snd (x₁, x₂)))
+    apply Filter.Tendsto.comp
+    apply Continuous.tendsto
+    exact continuous_snd
+    assumption
```

Comment:
I'll let you look and play with this suggestion rather than try to explain it, this is the perfect example for learning about our tactic `congrm`.
```suggestion
  rw [properSMul_iff_continuousSMul_ultrafilter_tendsto]
  refine and_congr_right fun hc ↦ ?_
  congrm ∀ 𝒰 x₁ x₂ hxx, ∃ g, ?_
  exact and_iff_right_of_imp fun hg ↦ tendsto_nhds_unique
    (hg.smul ((continuous_snd.tendsto _).comp hxx)) ((continuous_fst.tendsto _).comp hxx)
```

Reply:
This is quite an amazing shortcut!